lib/compat/wordpress-6.7/script-modules.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+/**
+ * Print data associated with Script Modules in Script tags.
+ *
+ * This embeds data in the page HTML so that it is available on page load.
+ *
+ * Data can be associated with a given Script Module by using the
+ * `script_module_data_{$module_id}` filter.
+ *
+ * The data for a given Script Module will be JSON serialized in a script tag with an ID
+ * like `wp-script-module-data-{$module_id}`.
+ */
+function gutenberg_print_script_module_data(): void {
+	$get_marked_for_enqueue = new ReflectionMethod( 'WP_Script_Modules', 'get_marked_for_enqueue' );
+	$get_marked_for_enqueue->setAccessible( true );
+	$get_import_map = new ReflectionMethod( 'WP_Script_Modules', 'get_import_map' );
+	$get_import_map->setAccessible( true );
+
+	$modules = array();
+	foreach ( array_keys( $get_marked_for_enqueue->invoke( wp_script_modules() ) ) as $id ) {
+		$modules[ $id ] = true;
+	}
+	foreach ( array_keys( $get_import_map->invoke( wp_script_modules() )['imports'] ) as $id ) {
+		$modules[ $id ] = true;
+	}
+
+	foreach ( array_keys( $modules ) as $module_id ) {
+		/**
+		 * Filters data associated with a given Script Module.
+		 *
+		 * Script Modules may require data that is required for initialization or is essential to
+		 * have immediately available on page load. These are suitable use cases for this data.
+		 *
+		 * This is best suited to a minimal set of data and is not intended to replace the REST API.
+		 *
+		 * If the filter returns no data (an empty array), nothing will be embedded in the page.
+		 *
+		 * The data for a given Script Module, if provided, will be JSON serialized in a script tag
+		 * with an ID like `wp-script-module-data-{$module_id}`.
+		 *
+		 * The dynamic portion of the hook name, `$module_id`, refers to the Script Module ID that
+		 * the data is associated with.
+		 *
+		 * @param array $data The data that should be associated with the array.
+		 */
+		$data = apply_filters( "script_module_data_{$module_id}", array() );
```

Comment:
Taking a deeper look, I believe this is always returning an empty array because [this code](https://github.com/WordPress/wordpress-develop/blob/a799101e46ef0a042814c189a6331c23a52a100c/src/wp-includes/interactivity-api/class-wp-interactivity-api.php#L311-L312) in core was added in 6.7, so it is not in 6.6.

That means that it is never entering the next conditional and it doesn't run `wp_print_inline_script_tag`.

Reply:
Close, but this was introduced in Gutenberg first, so this isn't entirely correct:

> That means that until 6.7, no developers should have used the add_filter(  "script_module_data_..." ) you shared in the example.

I would not add this compat code if it were not already in Gutenberg. But because the code _is_ in Gutenberg, the pre-existing Gutenberg implementation now becomes compat code for WordPress 6.7.

**Existing code of a new (experimental) feature in Gutenberg has transformed to compatibility code for WordPress 6.7.**

Reply:
Thanks for the context 🙂 I assumed it was introduced as an experiment, so it could be considered as "not added in Gutenberg."